libraries/AP_GPS/AP_GPS_SBP2.cpp

```diff
@@ -0,0 +1,494 @@
+            break;
+
+        case SBP2_detect_state::GET_MSG:
+            if (state.msg_type == SBP_HEARTBEAT_MSGTYPE && state.n_read < 4) {
+                *((uint8_t*)&(state.heartbeat_buff) + state.n_read) = data;
+            }
+            state.crc_so_far = crc16_ccitt(&data, 1, state.crc_so_far);
+            state.n_read += 1;
+            if (state.n_read >= state.msg_len) {
+                state.n_read = 0;
+                state.state = SBP2_detect_state::GET_CRC;
+            }
+            break;
+
+        case SBP2_detect_state::GET_CRC:
+            *((uint8_t*)&(state.crc) + state.n_read) = data;
+            state.n_read += 1;
+            if (state.n_read >= 2) {
+                state.state = SBP2_detect_state::WAITING;
+                if (state.crc == state.crc_so_far
+                        && state.msg_type == SBP_HEARTBEAT_MSGTYPE) {
+                    struct sbp_heartbeat_t* heartbeat = ((struct sbp_heartbeat_t*)state.heartbeat_buff);
+                    return heartbeat->protocol_major == 2;
+                }
+                return false;
+            }
+            break;
+
+        default:
+            state.state = SBP2_detect_state::WAITING;
+            break;
+    }
+    return false;
+}
+
+#if SBP_HW_LOGGING
+
+void
+AP_GPS_SBP2::logging_log_full_update()
+{
+
+    //CURRENTLY DEPRECATED. DO WE WANT TO RESSURRECT?
+
+    if (gps._DataFlash == nullptr || !gps._DataFlash->logging_started()) {
+      return;
+    }
+
+    //TODO: Expand with heartbeat info.
+    //TODO: Get rid of IAR NUM HYPO
+
+    struct log_SbpHealth pkt = {
+        LOG_PACKET_HEADER_INIT(LOG_MSG_SBPHEALTH),
+        time_us                    : AP_HAL::micros64(),
+        crc_error_counter          : crc_error_counter,
+        last_injected_data_ms      : last_injected_data_ms,
+        last_iar_num_hypotheses    : 0,
+    };
+    gps._DataFlash->WriteBlock(&pkt, sizeof(pkt));
+
+};
+
+void
+AP_GPS_SBP2::logging_log_raw_sbp(uint16_t msg_type,
+        uint16_t sender_id,
+        uint8_t msg_len,
+        uint8_t *msg_buff) {
+
+    if (gps._DataFlash == nullptr || !gps._DataFlash->logging_started()) {
+      return;
+    }
+
+    //MASK OUT MESSAGES WE DON'T WANT TO LOG
+    if (( ((uint16_t) gps._sbp_logmask) & msg_type) == 0) {
+        return;
+    }
+
+    uint64_t time_us = AP_HAL::micros64();
+
+    struct log_SbpRAW1 pkt = {
+        LOG_PACKET_HEADER_INIT(LOG_MSG_SBPRAW1),
+        time_us         : time_us,
+        msg_type        : msg_type,
+        sender_id       : sender_id,
+        msg_len         : msg_len,
+    };
+    memcpy(pkt.data1, msg_buff, MIN(msg_len,64));
+    gps._DataFlash->WriteBlock(&pkt, sizeof(pkt));
+
+    if (msg_len > 64) {
+
+        struct log_SbpRAW2 pkt2 = {
+            LOG_PACKET_HEADER_INIT(LOG_MSG_SBPRAW2),
+            time_us         : time_us,
+            msg_type        : msg_type,
+        };
+        memcpy(pkt2.data2, &msg_buff[64], msg_len - 64);
```

Comment:
IT's actually <= 192, since SBR1 stores the first 64 bytes (which most messages fit in) and SBR2 stores the other 192 bytes that enables storing a full 255 byte SBP message. Definitions in LogStructure.h:

#define LOG_SBP_STRUCTURES \
    { LOG_MSG_SBPHEALTH, sizeof(log_SbpHealth), \
      "SBPH", "QIII",   "TimeUS,CrcError,LastInject,IARhyp" }, \
    { LOG_MSG_SBPRAW1, sizeof(log_SbpRAW1), \
      "SBR1", "QHHBZ",      "TimeUS,msg_type,sender_id,msg_len,d1" }, \
    { LOG_MSG_SBPRAW2, sizeof(log_SbpRAW2), \
      "SBR2", "QHZZZ",      "TimeUS,msg_type,d2,d3,d4" }